### PR TITLE
Better organize migration guide

### PR DESCRIPTION
**Pull request by marcospereira**
## Purpose

The major sections for the migration guide are now:

- How to migrate
- API Changes
- Configuration changes
- Defaults changes
- Updated libraries
- Removed libraries
- Other important changes

This better groups sections that are related but weren't close to each other. It also adds complete code samples for the Http.Context migration guide and do some formatting changes there.

_review using [?w=1](https://github.com/playframework/playframework/pull/8843/files?w=1)_.

### CODE REVIEW

**Comment by mkurz** [comment]
Please have a look at #8842: There are some suggestions what we should add to the migration guide - plus `@Inject` annotations are missing.

File: documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md [hunk 1]
```diff
@@ -478,7 +542,7 @@ import play.mvc.Controller;
 
 public class FooController extends Controller {
     public Result info(Http.Request request) {
-        String message = request.flash().get("message");
+        String message = request.flash().getOptional("message").orElse("The default message");
```
Comment (ignasi35) [inline documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md]: Good catch.

**Comment by marcospereira** [comment]
@mkurz,

I've fixed the missing `@Inject` annotations.

**Comment by mkurz** [comment]
@marcospereira :+1: 

Maybe we can mention the warning 
```
method implicitXXX in object PlayMagicForJava is deprecated (since 2.7.0): See https://www.playframework.com/documentation/latest/JavaHttpContextMigration27
```
in [this section](https://www.playframework.com/documentation/2.7.0-RC3/JavaHttpContextMigration27#Some-template-tags-need-an-implicit-Request,-Messages-or-Lang-instance) as well? It looks like the reporter of #8842 was using the web-browser's search feature and searched for "PlayMagicForJava" but couldn't find anything... That's why he reported the issue.

File: documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md [hunk 2]
```diff
@@ -629,7 +631,46 @@ These tags will automatically use the implicit request passed to this template:
 @defaultpages.unauthorized
 ```
 
-Following tags need an implicit `Messages` instance to be present:
+So, if you have a view that use some of the tags above, for example if you have a file `app/views/names.scala.html` like below:
+
+```html
+@(names: List[String])(implicit request: Http.Request)
```
Comment (ignasi35) [inline documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md]: Doesn't this need the import? (`play.mvc.Http.Request`)

**Comment by marcospereira** [comment]
@mkurz please see the last two commits. :-)

**Comment by ignasi35** [comment]
Oh, these last few commits are great. I stumbled upon MagicJava a few times these past few days.

File: documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md (hunk 2 above)
Comment (marcospereira) [reply documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md]: No.

https://github.com/playframework/playframework/blob/ee30d34432cc26cd1ccfbe03911d7062104ba553/framework/src/build-link/src/main/java/play/TemplateImports.java#L34

https://github.com/playframework/playframework/blob/b307d04cde614323a76a40e80210472c4da82187/framework/src/sbt-plugin/src/main/scala/play/sbt/PlaySettings.scala#L38

https://github.com/playframework/playframework/blob/f5438b37a602ef804c65785ced972194bb29ad34/framework/src/sbt-plugin/src/main/scala/play/sbt/Play.scala#L76

:-)

Comment (ignasi35) [reply documentation/manual/releases/release27/migration27/JavaHttpContextMigration27.md]: I have to review a few PRs I sent recently, then. 😂

**Comment by marcospereira** [comment]
Backport to 2.7.x: 90f8a01b74622d2525ec2345bc43de8cd985767b

**Pull request merged**